### ISSUE TITLE
Keyboard inactive after running kmonad, terminal saying "connected" (intel Mac running Big Sur)

### ISSUE THREAD

**Issue by agemartin** [kmonad/kmonad#493]
I succesfully installed kmonad after upgrading to Big Sur from Mojave. However, the moment I try to run the keymap, the terminal answers "connected" and the keyboard stops responding. The only key which works is the infamous "key to the left of 1" (#106) which succesfully kills kmonad and brings the keyboard back to its unmanipulated life.

Any idea what does that mean and how I could possibly fix this?
Can I provide any info to clear this? (I am no IT guy and have very little idea of how these things work.)

I am simply running:

`sudo /Users/martin/kmonad/kmonad/.stack-work/install/x86_64-osx/b2cfacb6c2bda1d40c4a637bd7e1e157d529213cebf789b20f7d723e8369beb8/8.10.4/bin/kmonad /Users/martin/kmonad/keymap/agemartin-cz.kbd
`

BTW: Over a year using kmonad, I have to admit I appreciate this #106 disfunction, it makes switching kmonad off so easy and fast in case of any bugs or tweaking needed. I really see it as feature tbh.

Thanks for any help. KMonad changed my life for the better.

**Comment by MuhammedZakir** [comment]
1. Did you update Karabiner driver from kext to dext?
    - Remove kext and install dext.
2.  Did you rebuild KMonad for dext?

See the installation doc for more details. If you have already installed dext, check if the installed version matches the one mentioned in doc.

You can pass `--log debug` to `kmonad` cli (e.g. `sudo kmonad --log debug ~/kmonad/keymap/agemartin-cz.kbd`) to make KMonad print more information on what it's doing.

**Comment by agemartin** [comment]
thanks for the reply
1. I did not remove kext. I have not found in the documentation any info on removing the kext. The dext was installed via karabiner elements, however the version was higher so I have built the `1.15.0` version manually using the steps decribed in the installation doc. 
2. Yes, I did a clean clone as well.

I tried the --log debug, resulted in `invalid option '--log'`

MacBook-Pro:~ martin$ sudo /Users/martin/kmonad/kmonad/.stack-work/install/x86_64-osx/b2cfacb6c2bda1d40c4a637bd7e1e157d529213cebf789b20f7d723e8369beb8/8.10.4/bin/kmonad --log debug /Users/martin/kmonad/keymap/agemartin-cz.kbd
Invalid option `--log'

Usage: kmonad FILE [-d|--dry-run] [-l|--log-level Log level] 
              [-w|--start-delay ARG] [-c|--allow-cmd] [-f|--fallthrough] 
              [-t|--init STRING] [-s|--cmp-seq BUTTON] [-o|--output OTOKEN] 
              [-i|--input ITOKEN] [-V|--version]
  Start KMonad
MacBook-Pro:~ martin$

**Comment by MuhammedZakir** [comment]
Sorry, It's `--log-level`.

> 1. I did not remove kext. I have not found in the documentation any info on removing the kext.

When you update Karabiner-Elements, it should be removed IIRC. I told to remove it in case it causes some problem.

> The dext was installed via karabiner elements, however the version was higher (I tested it, it did not work, as described in other issues here)

We haven't supported new versions (yet).

> so I have built the `1.15.0` version manually using the steps described in the installation doc.

Built? Or did you mean that you installed using the built package provided in the karabiner-dext's repo? 

> 2. Yes, I did a clean clone as well.

I meant, did you pass the `dext` flag or `kext` flag when building KMonad?

**Comment by agemartin** [comment]
> Built? Or did you mean that you installed using the built package provided in the karabiner-dext's repo?

Sorry, I have INSTALLED the dext as described in the installation doc - in the "Install the already build and signed dext package" part, ie using.

> When you update Karabiner-Elements, it should be removed IIRC. I told to remove it in case it causes some problem.

So I guess no need to remove it manually? If so, how can this be done?

> I meant, did you pass the dext flag or kext flag when building KMonad?

I used the dext flag.

Generally I believe I followed correctly the installation steps as described in the doc. The installation seemed to have been completed sucesfully.

> It's --log-level.

Great, here we go. As long as I can understand the messages, all events are correctly translated/manipulated according to my keymap. Apart from the terminal where the messages appear upon pressing/releasing keys, the keyboard is not "working" - ie I cannot write, also no shortcuts work.

```
Opening KeySink
Initializing Mac key sink
Opening KeySource
Opening IOKit devices
Launching process: emitter_proc
connected

--------------------------------------------------------------------------------
Received event: Press <w>
Running hooks
Block level set to: 1
Registering untimed hook: 2
Registering 200ms hook: 1

--------------------------------------------------------------------------------
Received event: Release <w>
Running hooks
Emitting: Press <g>
Unblocking input stream, no stored events
Emitting: Release <g>
Tried cancelling expired hook: 1

--------------------------------------------------------------------------------
Received event: Press <d>
Running hooks
Block level set to: 1
Registering untimed hook: 4
Registering 150ms hook: 3

--------------------------------------------------------------------------------
Received event: Release <d>
Running hooks
Emitting: Press <r>
Unblocking input stream, no stored events
Emitting: Release <r>
Tried cancelling expired hook: 3


--------------------------------------------------------------------------------
Received event: Press <spc>
Running hooks
Block level set to: 1
Registering untimed hook: 79
Registering 160ms hook: 78
Cancelling hook: 78
Pushed layer to stack: lmetlay
 1. lmetlay
Base-layer: default

Registering untimed hook: 80
Emitting: Press <met>
Unblocking input stream, no stored events

--------------------------------------------------------------------------------
Received event: Release <spc>
Running hooks
Popped layer from stack: lmetlay
Base-layer: default

Emitting: Release <met>

--------------------------------------------------------------------------------
Received event: Press <met>
Running hooks
Block level set to: 1
Registering untimed hook: 82
Registering 200ms hook: 81
Cancelling hook: 81
Pushed layer to stack: lmetlay
 1. lmetlay
Base-layer: default

Registering untimed hook: 83
Emitting: Press <met>
Unblocking input stream, no stored events

--------------------------------------------------------------------------------
Received event: Press <tab>
Running hooks
Registering untimed hook: 84
Registering untimed hook: 85
Registering untimed hook: 86
Emitting: Press <tab>

--------------------------------------------------------------------------------
Received event: Release <tab>
Running hooks
Registering untimed hook: 87
Registering untimed hook: 88
Emitting: Release <tab>

--------------------------------------------------------------------------------
Received event: Release <met>
Running hooks
Popped layer from stack: lmetlay
Base-layer: default

Emitting: Release <met>


```

**Comment by MuhammedZakir** [comment]
Did you enabled input monitoring for your terminal in System Preferences? [1] Also, is KE running?

Btw, what's your macOS version?

> So I guess no need to remove it manually? If so, how can this be done?

Probably not. See if those files exist -- `/Library/Application Support/org.pqrs/Karabiner-VirtualHIDDevice` [2]. Don't remove `org.pqrs` folder, dext is installed there.

[1] https://github.com/kmonad/kmonad/blob/master/doc/installation.md#giving-kmonad-additional-permissions
[2] https://github.com/pqrs-org/Karabiner-VirtualHIDDevice/blob/master/dist/uninstall.sh#L19

**Comment by agemartin** [comment]
> Did you enabled input monitoring for your terminal in System Preferences? [1] 
Yes, terminal is ticked, Terminal and karabiner_graber and karabiner_observer

> Also, is KE running?
Nope, I have shut it down, Karabiner Elements is not running.

> Btw, what's your macOS version?

Big Sur 11.6 (20G165)

> See if those files exist -- /Library/Application Support/org.pqrs/Karabiner-VirtualHIDDevice [2]. Don't remove org.pqrs folder, dext is installed there.

there are three subfolders, Karabiner-DriverKit-VirtualHIDDevice, Karabiner-Elements and tmp.
I tried to run 
`2>/dev/null rmdir '/Library/Application Support/org.pqrs/Karabiner-VirtualHIDDevice'
` from [2], it seems to me it did not achieve anything. Any point deleting the Karabiner-VirtualHIDDevice subfolder manually? It goes without saying that I don't mind removing Karabiner Elements completely, I am not using it. 


[2] https://github.com/pqrs-org/Karabiner-VirtualHIDDevice/blob/master/dist/uninstall.sh#L19

**Comment by MuhammedZakir** [comment]
I will post here after doing some tests.

> there are three subfolders, Karabiner-DriverKit-VirtualHIDDevice, Karabiner-Elements and tmp.

It looks like kext has already been removed.

> I tried to run
`2>/dev/null rmdir '/Library/Application Support/org.pqrs/Karabiner-VirtualHIDDevice'` from [2], it seems to me it did not achieve anything. If I pass it without the 2> (??? don't understand that?), it answers
> `zsh: permission denied: /dev/null`

You need administrator/root privilege to delete files in certain directories and this is one of them. Use `sudo`.

`2>/dev/null` means "throw away all errors". See these 2 links for more details.

https://www.linuxshelltips.com/redirect-output-in-linux/
https://stackoverflow.com/questions/20494015/confusion-with-the-redirection-of-output-to-21/20494125#20494125

> Any point deleting the Karabiner-VirtualHIDDevice subfolder manually?

You can actually delete it using Finder too. You can use whichever method you're comfortable with.

> It goes without saying that I don't mind removing Karabiner Elements completely, I am not using it.

I keep it around, mostly for EventViewer and for some remappings which are impossible in KMonad (esp. mouse-related). It's not possible atm because of incompatible dext version. Updating it is in my todo list, so I can use both.

**Comment by agemartin** [comment]
> Can you post your defcfg part from your config here?

huh, I already thought we have it, since it says `output (kext)`. 

```

(defcfg
input (iokit-name "Apple Internal Keyboard / Trackpad")  ;; Macbook keyboard
output (kext)
fallthrough true
allow-cmd true
)
```

however - first shot - replacing it with `(dext)` resolves in parse error
```
kmonad: Parse error at 21:17:
   |
21 |                 output (dext)                                                                                                                                                                                
   |                 ^
unexpected 'o'
expecting ')'
```
can't find any alternative in the config tutorial file either



> I will post here after doing some tests.

thanks a lot, also for the other explanations, I really appreciate its

**Comment by MuhammedZakir** [comment]
To KMonad, using `dext` inside `output` is only supported in `keycode-refactor` branch. But, you don't need to change them because in config it's more like a placeholder (they mean the same thing) as the app only support either dext or kext and never both at the same time.

I won't be able to access Mac for a few days, so... try these.
1. You probably have done it, but still, have you "activated" the dext? The command is in the doc. You may have to use `forceActivate` instead of `activate` if using latter fails.
2. Disabe and re-enable input monitoring permission for your terminal.
3. Enable it for your `kmonad` binary.
4. Restart and do 2 & 3 again. (may seem weird, but it sometimes works)

**Comment by agemartin** [comment]
1. I activated it during the installation, tried the forceActivate now, it went through but did not affect anything in the end.
2. Tried
3. Tried
4. Tried

sadly, the behaviour is still the same. 

What I might do during the weekend is to do a clean installation of the OS and try to install kmonad just after that, prior the migration of apps and other data. Not sure if the migration would than overwrite the installtion? Don't have that much experience with the migration anyway, would be quiet awesome for me to avoid setting up and installing everthing on a clean copy of the OS though...

**Comment by agemartin** [comment]
I did a clean instalation of big sur and got kmonad running normally. 
I did not repeat the process of the data migration exactly the same way, I was creating a new user first, than migrated the old user stuff - and actually even made it to get kmonad running on the new user than (basically what was failing before).
So no idea what the problem was nor what the solution would have been, other than fresh instal etc..

Thank you very much for the help anyway!
